source/expressions.tex

```diff
@@ -5540,7 +5540,8 @@
 \pnum
 If the keyword \keyword{delete} in a \grammarterm{delete-expression}
 is not preceded by the unary \tcode{::} operator and the type of the operand is
-a pointer to a (possibly cv-qualified) class type \tcode{T}:
+a pointer to a (possibly cv-qualified) class type \tcode{T}
```

Comment:
For the first bullet, it's relevant to [class.dtor] p16, which is defined as
> At the point of definition of a virtual destructor (including an implicit definition), the **non-array deallocation function** is determined as if for the expression delete this appearing in a non-virtual destructor of the destructor's class (see [expr.delete]).   

However, the pointer may point to the initial element of such type of an array.

Reply:
Arrays cannot contain polymorphic data, thus it's fine to use the deallocation function from the definition of the virtual destructor, because the static and the dynamic type must be the same here.

Reply:
In [class.dtor] p16, it only specifies the approach about how to determine the **non-array deallocation function**. However, when an array object is being deleted, as per [expr.delete] p8 

> If a deallocation function is called, it is operator delete for a single-object delete expression or operator delete[] for an array delete expression.  

We should find `delete[]` for this purpose. 
````cpp
struct A{
    void operator delete[](void* ptr){
        std::cout<<"#1\n";
    }
    void operator delete(void* ptr){
      std::cout<<"#2\n";
    }
    virtual ~A(){}
};
int main(){
   auto ptr = new A[2][3];
   delete [] ptr;
}
````
`#1` invokes

Reply:
Ah, you're saying that [class.dtor] p16 only determines a non-array operator delete, but [expr.delete] p9.1 as changed by this pull request also wants to use some `operator delete[]` presumably found that way.

Reply:
Fixed.